pybliometrics/scopus/serial_search.py

```diff
@@ -0,0 +1,303 @@
+            https://dev.elsevier.com/guides/SerialTitleViews.htm.
+
+        Raises
+        ------
+        Scopus400Error
+            If provided value for a query key is invalid or for non-subscribers,
+            if the number of search results exceeds 5000.
+
+        ValueError
+            If provided query key is invalid.
+
+        Notes
+        -----
+        The directory for cached results is `{path}/{view}/{fname}`,
+        where `path` is specified in `~/.scopus/config.ini` and fname is
+        the md5-hashed version of `query` dict turned into string in format of
+        'key value' deliminated by '&'.
+        """
+        allowed_query_keys = set(['date', 'field', 'issn',
+                                  'oa', 'subj','subjCode',
+                                  'title'])
+        allowed_views = ('STANDARD', 'ENHANCED', 'CITESCORE')
+        if not set(query.keys()).issubset(allowed_query_keys):
+            raise ValueError('allowed query keys are: ' + 
+                             ', '.join(allowed_query_keys))
+        if view not in allowed_views:
+            raise ValueError('view parameter must be one of ' +
+                             ', '.join(allowed_views))
+        Search.__init__(self, query=query, api='SerialSearch',
+                        refresh=refresh, count=count, view=view)
+
+
+def _form_fields(results):
+    """Auxiliary function that returns tuple represiting union of keys
+    of all dictionaries inside search result list including keys of nested
+    dictionaries. Nested dictionaries correspond to links data, citescore
+    metric data from last available year, subject area data and other data
+    varying by year (e.g.  SNIP and SJR yearly data). Keys of periodic data
+    and citescore metrics data are treated by merging stat name and period with
+    underscore. If subject area data is present, keys corresponding to area
+    code, names and abbreviations are created. Keys of links data are
+    created corresponding to link names given by data.
+    """
+    special_keys = ['@_fa',
+                    'subject-area',
+                    'SNIPList',
+                    'SJRList',
+                    'citeScoreYearInfoList',
+                    'link',
+                    'yearly-data']
+    base_fields = set() # Subject area keys, links keys and keys with non-nested dictionaries
+    time_fields = set() # SJR, SNIP, citescore and yearly data keys
+    for i in results:
+        base_fields.update(set(key for key in i if key not in special_keys))
+        # Keys from dictionaries of yearly data
+        if i.get('yearly-data'):
+            yearly_fields = set()
+            for t in i['yearly-data'].get('info', []):
+                period_fields = set(key
+                                    + '_'
+                                    + str(t.get('@year'))
+                                    for key in t.keys()
+                                    if key not in ('@year', '@_fa'))
```

Comment:
This looks a bit too complicated. Why not `set(f"{k}_{t.get('@year')" for k in t if k not in ('@year', '@_fa'))`? Also, what happens if `t['@year']` returns `None`? Can it actually happen? If no, simply use `t['@year']`.